syntax = "proto3";

import "pipelines_and_context.proto";
import "util.proto";
import "google/protobuf/empty.proto";

option go_package = "./;grpcmodels";

package ficus;

service GrpcBackendBalancerService {
  rpc SetPipelinePartsToBackendsMap(GrpcPredefinedPipelinePartsToBackendsMap) returns (google.protobuf.Empty);
}

message GrpcPredefinedPipelinePartsToBackendsMap {
  repeated GrpcPipelinePartToBackends parts_to_backends = 1;
}

message GrpcPipelinePartToBackends {
  string part_name = 1;
  repeated string backends = 2;
}

service GrpcBackendService {
  rpc ExecutePipeline(GrpcProxyPipelineExecutionRequest) returns (stream GrpcPipelinePartExecutionResult);

  rpc GetContextValue(GrpcGetContextValueRequest) returns (GrpcGuid);
  rpc GetAllContextValues(GrpcGuid) returns (GrpcGetAllContextValuesResult);

  rpc DropExecutionResult(GrpcGuid) returns (google.protobuf.Empty);

  rpc GetBackendInfo(google.protobuf.Empty) returns (GrpcFicusBackendInfo);
}

message GrpcFicusBackendInfo {
  string name = 1;
  repeated GrpcPipelinePartDescriptor pipeline_parts = 2;
}

message GrpcPipelinePartDescriptor {
  string name = 1;
}

message GrpcGetContextValueRequest {
  GrpcGuid executionId = 2;
  GrpcContextKey key = 1;
}

message GrpcPipelineExecutionRequest {
  GrpcPipeline pipeline = 1;
  repeated GrpcContextKeyValue initialContext = 2;
}

message GrpcProxyPipelineExecutionRequest {
  GrpcPipeline pipeline = 1;
  repeated GrpcGuid contextValuesIds = 2;
}

message GrpcPipelinePartExecutionResult {
  oneof result {
    GrpcPipelineFinalResult finalResult = 1;
    GrpcPipelinePartResult pipelinePartResult = 2;
    GrpcPipelinePartLogMessage logMessage = 3;
  }
}

message GrpcPipelinePartLogMessage {
  string message = 1;
}

message GrpcPipelinePartResult {
  repeated GrpcContextValueWithKeyName contextValues = 1;
  GrpcGuid guid = 2;
}

message GrpcPipelineFinalResult {
  oneof executionResult {
    GrpcGuid success = 1;
    string error = 2;
  }
}

message GrpcGetAllContextValuesResult {
  repeated GrpcGuid context_values = 1;
}